compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -527,6 +527,54 @@ impl<'tcx> Body<'tcx> {
     pub fn is_custom_mir(&self) -> bool {
         self.injection_phase.is_some()
     }
+
+    /// *Must* be called once the full substitution for this body is known, to ensure that the body
+    /// is indeed fit for code generation or consumption more generally.
+    ///
+    /// Sadly there's no nice way to represent an "arbitrary normalizer", so we take one for
+    /// constants specifically. (`Option<GenericArgsRef>` could be used for that, but the fact
+    /// that `Instance::args_for_mir_body` is private and instead instance exposes normalization
+    /// functions makes it seem like exposing the generic args is not the intended strategy.)
+    ///
+    /// Also sadly, CTFE doesn't even know whether it runs on MIR that is already polymorphic or still monomorphic,
+    /// so we cannot just immediately ICE on TooGeneric.
+    ///
+    /// Returns Ok(()) if everything went fine, and `Err` if a problem occurred and got reported.
+    pub fn post_mono_checks(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+        normalize_const: impl Fn(ConstantKind<'tcx>) -> Result<ConstantKind<'tcx>, ErrorHandled>,
+    ) -> Result<(), ErrorHandled> {
+        // For now, the only thing we have to check is is to ensure that all the constants used in
+        // the body successfully evaluate.
+        for &const_ in &self.required_consts {
+            let uv = match normalize_const(const_.literal)? {
+                ConstantKind::Ty(ct) => match ct.kind() {
+                    ty::ConstKind::Unevaluated(uv) => uv.expand(),
+                    ty::ConstKind::Value(..) => {
+                        // These are already evaluated, so this cannot fail.
+                        return Ok(());
+                    }
+                    ty::ConstKind::Error(err) => {
+                        return Err(ErrorHandled::Reported(err.into(), const_.span));
+                    }
+                    ty::ConstKind::Param(..)
+                    | ty::ConstKind::Infer(..)
+                    | ty::ConstKind::Bound(..)
+                    | ty::ConstKind::Placeholder(..)
+                    | ty::ConstKind::Expr(..) => return Err(ErrorHandled::TooGeneric(const_.span)),
+                },
+                ConstantKind::Unevaluated(uv, _) => uv,
+                ConstantKind::Val(..) => return Ok(()),
```

Comment:
We have this kind of match in a bunch of places (here, interpret, codegen_ssa, codegen_cranelift). There doesn't happen to be some `eval` function we can call instead? There are `eval` methods on constants but they return a constant again, i.e. the return type doesn't reflect that anything was evaluated -- not quite what I was expecting from an `eval` function.

Reply:
Why do we need a new enum? For `ty::Const` it's a Valtree, for `mir::Const` its a `ConstValue`.

Reply:
`mir::Const` would then need to convert `Valtree` to `ConstValue` in case it starts out with `ty::Const`.

Reply:
Yeah, that's what I would expect to happen, indeed. Miri/const-eval already do that.

Reply:
This seems to actually have worked. :)